godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::ptr::drop_in_place;
+use std::{cell::RefCell, rc::Rc};
+
+use crate::engine::{Object, Script, ScriptLanguage};
+use crate::obj::Gd;
+use crate::sys;
+
+use super::meta::{MethodInfo, PropertyInfo};
+use super::{GString, StringName, Variant};
+
+/// Interface for Godot's `GDExtensionScriptInstance`, which cannot be directly constructed by an extension. Instead a type that implements this
+/// trait has to be passed to the [`create_script_instance`] function, which returns a [`sys::GDExtensionScriptInstancePtr`] pointer. This pointer
+/// can then be returned from [`IScriptExtension::instance_create`](crate::engine::IScriptExtension::instance_create).
+pub trait ScriptInstance {
+    /// Name of the new class the script implements.
+    fn class_name(&self) -> GString;
+
+    /// Property setter for Godots virtual dispatch system. The engine will call this function when it wants to change a property on the script.
+    fn set(&mut self, name: StringName, value: &Variant) -> bool;
+
+    /// Property getter for Godots virtual dispatch system. The engine will call this function when it wants to read a property on the script.
+    fn get(&self, name: StringName) -> Option<Variant>;
+
+    /// A list of all the properties a script exposes to the engine.
+    fn property_list(&self) -> Rc<Vec<PropertyInfo>>;
+
+    /// A list of all the methods a script exposes to the engine.
+    fn method_list(&self) -> Rc<Vec<MethodInfo>>;
+
+    /// Method invoker for Godots virtual dispatch system. The engine will call this function when it wants to call a method on the script.
+    ///
+    /// All method calls are taking a mutable reference of the script instance, as the engine does not differentiate between immutable and
+    /// mutable method calls like rust.
+    ///
+    /// It's important that the script does not cause a second call to this function while executing a method call. This would result in a panic.
+    fn call(
+        &mut self,
+        method: StringName,
+        args: &[&Variant],
+    ) -> Result<Variant, sys::GDExtensionCallErrorType>;
+
+    /// Identifies the script instance as a placeholder. Godot treats placeholder instances slightly different than normal instances.
+    fn is_placeholder(&self) -> bool;
+
+    /// Validation function for the engine to verify if the script exposes a certain method.
+    fn has_method(&self, method: StringName) -> bool;
+
+    /// Lets the engine get a reference to the script this instance was created for.
+    fn get_script(&self) -> Gd<Script>;
+
+    /// Lets the engine fetch the type of a particular property.
+    fn get_property_type(&self, name: StringName) -> sys::VariantType;
```

Comment:
`VariantType` should be re-exported in `godot::builtin`, maybe you can use that one.

Ideally there are no `sys::` symbols in public APIs; that crate is for implementation only.